### DISCUSSION TITLE
Upgrading from 2.2.x to a recent version like 2.9

### DISCUSSION THREAD

**Discussion by researchauq** [ckan/ckan#6969] [Q&A]
Inherited a very old CKAN install, and have been playing with various things to see if we could get things up to date. Hard to do much as there is very little documentation on a version this old.

But I'm not sure on updating; most questions have talked about moving one minor/patch version, not several.

So I'm wondering if we have to do ... 2.2->2.3->2.4->2.5->2.6->2.7->2.8->2.9, or if there are some/all we could skip?

I'm still a bit early on in learning CKAN, so I'm not sure what might be needed. I'm thinking there are DB upgrades, but maybe config file changes that I have to be mindful of?

Crazy q: Would it be possible to export the DB on 2.2 and import it onto a fresh install of 2.9? Guessing no.

Any advice or pitfall warnings?

**Comment by wardi** [comment]
you can dump and load many objects using the `ckanapi` CLI tool, including datasets, groups, organizations, users, and related items, but not things like revisions and activities so your history on the new site would be blank.

**Reply by researchauq** [reply]
Are you talking about export on 2.2 and straight import on 2.9?

(Would it be possible to retain activities and revisions if we did the minor by minor version upgrade?)

For the "db upgrades" (if needed) ... I think you need to install the new software, and then run the db-upgrade on the current DB ... so I think if I install a new VM with a new version, I have to restore the old-version dumped DB and then run the db-upgrade? (I was actually a bit confused on this, but writing it out it looks right if my assumptions are correct).

**Comment by amercader** [comment]
When migrating a site you need to:
1. Upgrade your extensions. For this case I generally don't bother with going one version at a time and jump straight to the latest version, assuming you have a decent test coverage. I'd start migrating the tests to pytest (using the `pytest-ckan` package) and the new testing conventions, adding any extra tests needed to cover all the basic functionality. Once these pass, I'd start writing the new version of the extension against CKAN 2.9 (there are tips in this [wiki page](https://github.com/ckan/ckan/wiki/Python-3-migration-guide-for-extensions). I've never migrated such an old version like yours so you might need intermediate steps (perhaps you want to migrate to latest CKAN with py2 first and then py3)
2. Upgrade your database. Each CKAN version includes changes in the database tables. You have two options
   * Run the db migrations. I'm not sure if a direct migration from 2.2 to 2.9 will work but there's no harm in trying. Just boot up a fresh 2.9 instance, point it to a copy of the current 2.2 db and run `ckan db upgrade` and see what happens. We switched from sqlalchemy-migrate to alembic for migrations in 2.9, I'm not sure what effect this will have, if any
   * Export your current objects and recreate them via the API using ckanapi as @wardi suggested, This is probably the cleanest method, but you'll lose revisions/activities information

**Reply by researchauq** [reply]
First, let me say thanks to both of you for the speedy replies; this is so helpful.

For the extensions, I'm running this command:
/usr/lib/ckan/default/bin/paster --plugin=ckan plugin-info -c /etc/ckan/default/production.ini

Which spits out what looks like a bunch of Python-related programming bits (plugin name, what it implements), but then it stops with a Python traceback. This is how I inherited it, I'm just wondering if there is any other way to check what the extensions are?

I'm definitely game to try to get a workable recent version, and the fiddle a bit to see if we can do anything on the revisions and activities.
I sheepishly admit I only recently understood the documents were stored outside of the database; for some odd reason I thought everything was in PostgreSQL and that the outside files were caches, but I'm learning.

Part of the issue I've had is that all of the docs are recent in the Write the Docs space, with the oldest version still available being 2.5.9 that I could find. I now found this is all in Git so have some help there too; but even that is referencing ancient versions of OSes (Ubuntu 12 or 14).
I'll try what you both have suggested and add it to the thread; thanks again.

**Comment by duttonw** [comment]
Your up for a large battle if you have any custom theme or plugins in getting updated. Since you have not provided the public url, we can't really point you in directions for plugin improvements. 

Qld Gov has been doing dramatic improvments to get to 2.8 from 2.6-7 and now just done major works on a large suite of plugins to be 2.9.x compatible with swapping out the MVC layer (plyon,flask, duel living to only one being used in 2.9).

You can see quite a bit of our work in this diff plus many other plugins we 'forked' to make them compatible for our system (plus we had external's do 'customisations' on top that were not done in a fashion that could be folded back upstream with that feature being able to be turned off etc). 

https://github.com/qld-gov-au/ckanext-validation/compare/v0.0.7...v0.0.8-qgov.5
https://github.com/qld-gov-au/ckanext-report/compare/0.2...0.3.1-qgov.1
https://github.com/qld-gov-au/ckanext-datarequests/compare/2.1.1-qgov...2.2.1-qgov.3
